documentation/intermediate-ca.md

```diff
@@ -0,0 +1,199 @@
+    Files /dev/fd/63 and /dev/fd/62 differ
+
+##### Solution
+
+Remediating this issue is tricky because if the wrong CRL has been copied onto the Puppet Server then it's possible for certificate revocation information to be permanently lost. The best way to remediate this issue is to remove the CRL entirely, issue a dummy certificate, and then revoke/clean it to generate a new CRL.
+
+    rm $(puppet agent --configprint hostcrl) $(puppet master --configprint cacrl)
+    puppet cert generate revokeme
+    puppet cert clean revokeme
+
+    [root@pe-master ~]# puppet cert generate revokeme
+    Notice: revokeme has a waiting certificate request
+    Notice: Signed certificate request for revokeme
+    Notice: Removing file Puppet::SSL::CertificateRequest revokeme at '/etc/puppetlabs/puppet/ssl/ca/requests/revokeme.pem'
+    Notice: Removing file Puppet::SSL::CertificateRequest revokeme at '/etc/puppetlabs/puppet/ssl/certificate_requests/revokeme.pem'
+    [root@pe-master ~]# puppet cert clean revokeme
+    Notice: Revoked certificate with serial 7
+    Notice: Removing file Puppet::SSL::Certificate revokeme at '/etc/puppetlabs/puppet/ssl/ca/signed/revokeme.pem'
+    Notice: Removing file Puppet::SSL::Certificate revokeme at '/etc/puppetlabs/puppet/ssl/certs/revokeme.pem'
+    Notice: Removing file Puppet::SSL::Key revokeme at '/etc/puppetlabs/puppet/ssl/private_keys/revokeme.pem'
+
+#### The Puppet CA cannot generate certificates when the CA bundle has been installed
+
+##### Error message
+
+    [root@pe- ~]# puppet cert generate failtosign
+    Error: The certificate retrieved from the master does not match the agent's private key.
+    Certificate fingerprint: 10:0C:DF:B0:41:91:46:BF:1B:A8:F4:F5:44:88:1D:99:F3:B3:AE:3C:3A:E4:24:66:FB:50:CB:4A:20:FE:4F:6D
+    To fix this, remove the certificate from both the master and the agent and then start a puppet run, which will automatically regenerate a certificate.
+    On the master:
+      puppet cert clean pe-20173nightly-master.puppetdebug.vlan
+    On the agent:
+      1a. On most platforms: find /etc/puppetlabs/puppet/ssl -name pe-20173nightly-master.puppetdebug.vlan.pem -delete
+      1b. On Windows: del "\etc\puppetlabs\puppet\ssl\certs\pe-20173nightly-master.puppetdebug.vlan.pem" /f
+      2. puppet agent -t
+
+##### Cause
+
+The `puppet cert generate` command should use the first certificate in the `cacert` file, but unintentionally uses the `localcacert` file. If the first certificate in this file is not the Puppet intermediate CA certificate Puppet will load the CA key and a mismatched CA certificate, and will fail with a misleading error message. The logic for verifying that the key and certificate match is shared with the agent key/certificate verification and will report that the agent key and certificate are mismatched when the CA certificate and key are mismatched.
```

Comment:
> The logic for verifying that the key and certificate match is shared with the agent key/certificate verification and will report that the agent key and certificate are mismatched when the CA certificate and key are mismatched.

Can you clarify this? Specifically:

> The logic for verifying that the key and certificate match is shared with the agent key/certificate verification

What is sharing the logic, and to what is that logic being shared?

> and will report that the agent key and certificate are mismatched when the CA certificate and key are mismatched.

I feel like I'm missing some nuance here. It reports the key and certificate are mismatched, when the key and certificate are mismatched?

In other words, what is the difference between:

> the agent key and certificate are mismatched

and

> the CA certificate and key are mismatched